Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/test/TestCertificateQRCodeExtractor.kt

```diff
@@ -1,25 +1,127 @@
+class TestCertificateQRCodeExtractor @Inject constructor(
+    private val coseDecoder: HealthCertificateCOSEDecoder,
+    private val headerParser: HealthCertificateHeaderParser,
+    private val bodyParser: TestCertificateDccParser,
+) {
 
     /**
-     * May throw an **[InvalidHealthCertificateException]**
+     * May throw an **[InvalidTestCertificateException]**
      */
     fun extract(
         decryptionKey: ByteArray,
-        encryptedCoseComponents: ByteArray,
+        rawCoseObjectEncrypted: ByteArray,
     ): TestCertificateQRCode {
-        throw NotImplementedError()
+        val rawCoseObject = rawCoseObjectEncrypted.decrypt(decryptionKey)
+        return TestCertificateQRCode(
+            testCertificateData = rawCoseObject.decode(),
+            qrCode = rawCoseObject.encode()
+        )
     }
 
     /**
-     * May throw an **[InvalidHealthCertificateException]**
+     * May throw an **[InvalidTestCertificateException]**
      */
-    fun extract(qrCode: String): TestCertificateQRCode {
-        throw NotImplementedError()
+    fun extract(qrCode: String) = TestCertificateQRCode(
+        testCertificateData = qrCode.extract(),
+        qrCode = qrCode
+    )
+
+    private fun RawCOSEObject.decrypt(decryptionKey: ByteArray): RawCOSEObject = try {
+        coseDecoder.decryptMessage(
+            input = this,
+            decryptionKey = decryptionKey
+        )
+    } catch (e: InvalidHealthCertificateException) {
+        throw InvalidTestCertificateException(e.errorCode)
+    } catch (e: Throwable) {
+        throw InvalidTestCertificateException(HC_COSE_MESSAGE_INVALID)
+    }
+
+    private fun String.extract(): TestCertificateData =
+        removePrefix(PREFIX)
+            .decodeBase45()
+            .decompress()
+            .decode()
+
+    private fun RawCOSEObject.encode(): String {
+        return PREFIX + compress().encodeBase45()
+    }
+
+    private fun RawCOSEObject.decode(): TestCertificateData = try {
+        coseDecoder.decode(this).parse()
+    } catch (e: InvalidHealthCertificateException) {
+        throw InvalidTestCertificateException(e.errorCode)
+    } catch (e: Throwable) {
+        throw InvalidTestCertificateException(HC_COSE_MESSAGE_INVALID)
+    }
+
+    private fun CBORObject.parse(): TestCertificateData = try {
+        TestCertificateData(
+            header = headerParser.parse(this),
+            certificate = bodyParser.parse(this)
+        ).also {
+            Timber.v("Parsed test certificate for %s", it.certificate.nameData.givenNameStandardized)
+        }
+    } catch (e: InvalidHealthCertificateException) {
+        throw InvalidTestCertificateException(e.errorCode)
+    } catch (e: Throwable) {
+        throw InvalidTestCertificateException(HC_CBOR_DECODING_FAILED)
+    }
+
+    private fun String.decodeBase45(): ByteArray = try {
+        Base45Decoder.decode(this)
+    } catch (e: Throwable) {
+        Timber.e(e)
```

Comment:
can you please some message to the logs of `Timber.e()` to differentiate between them 👇🏼 ?

Reply:
What do you mean exactly?

Reply:
I mean: we have many  `Timber.e(e)` in the file, can we add at least the error code as message to be to search in the code and map them easily? such as 
```
Timber.e(e,"DECODING_ERRORCODE")

Timber.e(e,"ENCODING_ERRORCODE")
```